spec/driver_spec.cr

```diff
@@ -49,20 +49,20 @@ describe PlaceOS::Driver::DriverManager do
 
     {{PlaceOS::Driver::CONCRETE_DRIVERS.values.first[1]}}.functions.should eq(%({
       "switch_input":{
-        "input":["String"]
+        "input":[{"type":"string","enum":["HDMI","DisplayPort","HDBaseT"]}]
       },
       "add":{
-        "a":["Int32"],
-        "b":["Int32"]
+        "a":[{"type":"integer"}],
+        "b":[{"type":"integer"}]
       },
       "splat_add":{},
       "perform_task":{
-        "name":["String | Int32"]
+        "name":[{"anyOf":[{"type":"integer"},{"type":"string"}]}]
       },
       "error_task":{},
       "future_add":{
-        "a":["Int32"],
-        "b":["Int32", 200]
+        "a":[{"type":"integer"}],
+        "b":[{"type":"integer"}, 200]
```

Comment:
JSON schema provides a [`default` annotation](https://json-schema.org/understanding-json-schema/reference/generic.html#id2). With this now less opaque, could that be used in place of the array?

Reply:
although removing the array is a breaking change: https://github.com/PlaceOS/driver/blob/master/src/placeos-driver/driver_model.cr#L34

Reply:
I think even removing the default value is a breaking change: https://github.com/PlaceOS/driver/blob/master/src/placeos-driver/proxy/driver.cr#L144
but agree that we should get the default into the schema

Reply:
(cross module function calls are constructed in a similar way to how compilers deal with default values - i.e. there is no such thing as defaults on the compiled function, calling functions have to add the defaults)

Reply:
so having the default in the array is still more performant as otherwise we'd need to parse the json schema to extract the default value when making cross module calls

Reply:
The lexer will have already run across the full object. If other attributes are not needed, `default` can be targeted directly by specifying it as the `root`. This is possible either [directly](https://github.com/crystal-lang/crystal/blob/dd40a2442/src/json/from_json.cr#L16-L29) or by defining a partial object with a `JSON::Serialiable` annotation with the `root` param.

Reply:
agreed, but it is a breaking change - at the very least would require every driver to be recompiled on existing systems and all 3rd party repos would have to be updated to the latest driver code

Reply:
i think it wouldn't be too hard on the backoffice side.
the difference comes down to a comparison between an array vs an object.

similarly, we could aim to deprecate rather than outright break the current interface.

i think a cleaner approach in the long term is worth the effort.

Reply:
OK we have a backwards compatible implementation

Reply:
with a clean new interface

Reply:
we can completely deprecate it in a few months time